### ISSUE TITLE
Do I have to compile a TF.keras model that's part of a compiled model in order for its added losses to be taken into account during training?

### ISSUE THREAD

**Issue by kristofgiber** [tensorflow/tensorflow#32421]
I have already tried to get an answer [here](https://stackoverflow.com/questions/57851525) to this question because I cannot figure it out based on the documentation:

I have a VAE tf.keras model and it consists of an encoder and decoder model. I want to regularize the encoder and the decoder with different custom terms (using add_loss()) while training VAE (using VAE.fit()).

In pseudocode:

```
encoder = model(input, encoding)
decoder = model(encoding, output)
encoder.add_loss(custom_loss_1)
decoder.add_loss(custom_loss_2)
vae = model(input, output)
vae.compile(optimizer, standard_keras_loss_function)
vae.fit(args)
```

Note that I can't add the custom losses to the wrapper model (vae) because I want them to be backpropagated only in their respective submodels (encoder's loss shouldnt affect decoder's gradients and vice versa). I also can't fit encoder and decoder separately, VAEs benefit from jointly training the encoder and decoder.

I'm facing two challenges:

Although I see no error during training, I have no easy way to heck if encoder and decoder losses are taken into account when fitting bar wrapper model. Do I need to compile encoder and decoder for the added custom losses to be taken into account?

I don't want the 'standard_keras_loss' (binary crossentropy) to be backpropagated beyond the decoder, it should only affect the decoder's gradients. However, vae.compile() method requires me to add at least one loss to the VAE model. Thus, the only way to prevent this loss from backpropagating to encoder is to try and add a stop_gradient lambda layer after the encoder but this would disconnect my encoder from my decoder (in fact I'm getting disconnected graph error wherever I try to add the stop_gradient layer).

Is there a TF.keras way to do this without having to write custom training loop? I'm trying to keep a clean code.

**Comment by gowthamkpr** [comment]
@kristofgiber Can you please take a look at this [answer](https://stackoverflow.com/questions/54400971/keras-clean-implementation-for-multiple-outputs-and-custom-loss-functions) and let me know if it helps. Thanks!

**Comment by gowthamkpr** [comment]
Closing this issue as it has been inactive for more than 5 days. Please add additional comments and we can open the issue again. Thanks!
